mlflow/langchain/_langchain_autolog.py

```diff
@@ -0,0 +1,296 @@
+        callbacks.append(mlflow_callback)
+        kwargs["callbacks"] = callbacks
+        return args, kwargs
+
+
+def _runnable_with_retriever(model):
+    from langchain.schema import BaseRetriever
+
+    with contextlib.suppress(ImportError):
+        from langchain.schema.runnable import RunnableBranch, RunnableParallel, RunnableSequence
+        from langchain.schema.runnable.passthrough import RunnableAssign
+
+        if isinstance(model, RunnableBranch):
+            return any(_runnable_with_retriever(runnable) for _, runnable in model.branches)
+
+        if isinstance(model, RunnableParallel):
+            return any(_runnable_with_retriever(runnable) for runnable in model.steps.values())
+
+        if isinstance(model, RunnableSequence):
+            return any(_runnable_with_retriever(runnable) for runnable in model.steps)
+
+        if isinstance(model, RunnableAssign):
+            return _runnable_with_retriever(model.mapper)
+
+    return isinstance(model, BaseRetriever)
+
+
+def _chain_with_retriever(model):
+    with contextlib.suppress(ImportError):
+        from langchain.chains import RetrievalQA
+
+        return isinstance(model, RetrievalQA)
+    return False
+
+
+def patched_inference(func_name, original, self, *args, **kwargs):
+    """
+    A patched implementation of langchain models inference process which enables logging the
+    following parameters, metrics and artifacts:
+
+    - model
+    - metrics
+    - data
+
+    We patch either `invoke` or `__call__` function for different models
+    based on their usage.
+    """
+
+    import langchain
+
+    # import from langchain_community for test purpose
+    from langchain_community.callbacks import MlflowCallbackHandler
+
+    class _MLflowLangchainCallback(MlflowCallbackHandler, metaclass=ExceptionSafeAbstractClass):
+        """
+        Callback for auto-logging metrics and parameters.
+        We need to inherit ExceptionSafeAbstractClass to avoid invalid new
+        input arguments added to original function call.
+        """
+
+        def __init__(self, *args, **kwargs):
+            super().__init__(*args, **kwargs)
+
+    _lc_version = Version(langchain.__version__)
+    if not MIN_REQ_VERSION <= _lc_version <= MAX_REQ_VERSION:
+        warnings.warn(
+            "Autologging is known to be compatible with langchain versions between "
+            f"{MIN_REQ_VERSION} and {MAX_REQ_VERSION} and may not succeed with packages "
+            "outside this range."
+        )
+
+    run_id = self.run_id if hasattr(self, "run_id") else None
+    if (active_run := mlflow.active_run()) and run_id is None:
+        run_id = active_run.info.run_id
+    # TODO: test adding callbacks works
+    mlflow_callback = _MLflowLangchainCallback(
+        tracking_uri=mlflow.get_tracking_uri(),
+        run_id=run_id,
+    )
+    args, kwargs = _inject_mlflow_callback(func_name, mlflow_callback, args, kwargs)
+    with disable_autologging():
```

Comment:
do we disable autologging when running the inference function because the logging is handled by the `mlflow_callback` injection?

Reply:
It's to avoid patch internal functions that's called within the inference function, for runnable sequence as an example, it invokes each step's inference function, if we don't disable autologging then each step would log metrics/models and it messes up